substrate/frame/staking/src/ledger.rs

```diff
@@ -201,6 +221,30 @@ impl<T: Config> StakingLedger<T> {
 		}
 	}
 
+	/// Sets the ledger controller to its stash.
+	pub(crate) fn set_controller_to_stash(self) -> Result<(), Error<T>> {
+		let controller = self.controller.as_ref()
+            .defensive_proof("Ledger's controller field didn't exist. The controller should have been fetched using StakingLedger.")
+            .ok_or(Error::<T>::NotController)?;
+
+		ensure!(self.stash != *controller, Error::<T>::AlreadyPaired);
+
+		// check if the ledger's stash is a controller of another ledger.
+		if let Some(bonded_ledger) = Ledger::<T>::get(&self.stash) {
+			// there is a ledger bonded by the stash. In this case, the stash of the bonded ledger
+			// should be the same as the ledger's stash. Otherwise fail to prevent data
+			// inconsistencies. See <https://github.com/paritytech/polkadot-sdk/pull/3639> for more
+			// details.
+			ensure!(bonded_ledger.stash == self.stash, Error::<T>::BadState);
+		}
```

Comment:
Is this check redundant if ledger is read from our custom getter?

Reply:
No they are not the same. The custom getter checks for:

- given a `Bonded(stash, controller)`, the ledger of the controller is the correct one (i.e. `Ledger(controller).stash == stash`). So we allow double (sane) bonded ledgers to be fetched and mutated. This check is more lenient, since it's not necessarily wrong to mutate double bonded ledgers.

This check, on the other hand, is checking if the stash of the ledger is a controller of *another* ledger. In this case, we want to prevent the double bonded ledgers from running the `set_controller` otherwise it may get into a bad state. This check is stricter.